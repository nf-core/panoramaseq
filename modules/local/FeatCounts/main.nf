//custom_featurecounts

process FEATURECOUNTS_CUSTOM {
    tag "${meta.id}"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://depot.galaxyproject.org/singularity/subread:2.0.6--he4a0461_2' :
        'biocontainers/subread:2.0.6--he4a0461_2' }"

    //
    // Now “bam” is a single file, not a list:
    //
    input:
    tuple val(meta), path(bam), path(annotation)

    //
    // “-R BAM” outputs "<prefix>.featureCounts.bam", so we emit that file.
    //
    output:
        tuple val(meta), path("*.featureCounts.tsv"), emit: counts
        tuple val(meta), path("*.featureCounts.tsv.summary"), emit: summary
        tuple val(meta), path("*.featureCounts.bam"), emit: annotated_bam
        path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"
    def feature_type = task.ext.feature_type ?: 'exon'
    def attribute_type = task.ext.attribute_type ?: 'gene_id'
    def strandedness = task.ext.strandedness ?: '0'  // 0=unstranded, 1=stranded, 2=reverse-stranded
    def paired_end = meta.single_end ? '' : '-p'  // Add -p flag for paired-end data
    """
    featureCounts \\
        -s ${strandedness} \\
        ${paired_end} \\
        -T ${task.cpus} \\
        -t ${feature_type} \\
        -g ${attribute_type} \\
        -a ${annotation} \\
        -o ${prefix}.featureCounts.tsv \\
        -R BAM \\
        ${args} \\
        ${bam}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        subread: \$( echo \$(featureCounts -v 2>&1) | sed -e "s/featureCounts v//g")
END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}.featureCounts.tsv
    touch ${prefix}.featureCounts.tsv.summary
    touch ${prefix}.featureCounts.bam

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        subread: \$( echo \$(featureCounts -v 2>&1) | sed -e "s/featureCounts v//g")
END_VERSIONS
    """
}
