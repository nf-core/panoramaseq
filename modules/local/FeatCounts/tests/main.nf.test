nextflow_process {

    name "Test Process FEATURECOUNTS_CUSTOM"
    script "../main.nf"
    process "FEATURECOUNTS_CUSTOM"
    config "./nextflow.config"
    tag "modules"
    tag "modules_local"
    tag "featurecounts_custom"

    test("sarscov2 [bam] - single end") {

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/bam/test.single_end.bam", checkIfExists: true),
                    file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.counts != null },
                { assert process.out.summary != null },
                { assert process.out.annotated_bam != null },
                { assert snapshot(process.out.versions).match("single_end_versions") },
                { assert file(process.out.counts.get(0).get(1)).exists() },
                { assert file(process.out.summary.get(0).get(1)).exists() },
                { assert file(process.out.annotated_bam.get(0).get(1)).exists() }
            )
        }
    }

    test("sarscov2 [bam] - paired end") {

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:false ], // meta map
                    file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam", checkIfExists: true),
                    file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert process.out.counts != null },
                { assert process.out.summary != null },
                { assert process.out.annotated_bam != null },
                { assert snapshot(process.out.versions).match("paired_end_versions") },
                { assert file(process.out.counts.get(0).get(1)).exists() },
                { assert file(process.out.summary.get(0).get(1)).exists() },
                { assert file(process.out.annotated_bam.get(0).get(1)).exists() },
                { assert process.out.counts.get(0).get(1) ==~ ".*\\.featureCounts\\.tsv" },
                { assert process.out.summary.get(0).get(1) ==~ ".*\\.featureCounts\\.tsv\\.summary" },
                { assert process.out.annotated_bam.get(0).get(1) ==~ ".*\\.featureCounts\\.bam" }
            )
        }
    }

    test("sarscov2 [bam] - stub") {

        options "-stub"

        when {
            params {
                outdir   = "$outputDir"
            }
            process {
                """
                input[0] = [
                    [ id:'test', single_end:true ], // meta map
                    file(params.modules_testdata_base_path + "genomics/sarscov2/illumina/bam/test.single_end.bam", checkIfExists: true),
                    file(params.modules_testdata_base_path + "genomics/sarscov2/genome/genome.gtf", checkIfExists: true)
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
