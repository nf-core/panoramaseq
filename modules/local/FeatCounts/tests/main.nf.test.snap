{
    "paired_end_versions": {
        "content": [
            [
                "versions.yml:md5,c0b9bed7072a083075975b75ac9036a6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-06-24T15:25:00.874973924"
    },
    "sarscov2 [bam] - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.featureCounts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.featureCounts.tsv.summary:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.featureCounts.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    "versions.yml:md5,c0b9bed7072a083075975b75ac9036a6"
                ],
                "annotated_bam": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.featureCounts.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "counts": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.featureCounts.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "summary": [
                    [
                        {
                            "id": "test",
                            "single_end": true
                        },
                        "test.featureCounts.tsv.summary:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c0b9bed7072a083075975b75ac9036a6"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-06-24T15:11:33.661958876"
    },
    "single_end_versions": {
        "content": [
            [
                "versions.yml:md5,c0b9bed7072a083075975b75ac9036a6"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.2"
        },
        "timestamp": "2025-06-24T15:24:54.318535623"
    }
}